[PyTorch] List::operator[] can return const ref for Tensor & string

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50083 [PyTorch] List::operator[] can return const ref for Tensor & string**

This should supercede D21966183
(https://github.com/pytorch/pytorch/pull/39763) as the way to get fast
access to the contents of a `torch::List`.

Differential Revision: [D25776232](https://our.internmc.facebook.com/intern/diff/D25776232/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25776232/)!

swolchok committed:
[PyTorch] List::operator[] can return const ref for Tensor & string

This should supercede D21966183
(https://github.com/pytorch/pytorch/pull/39763) as the way to get fast
access to the contents of a `torch::List`.

Differential Revision: [D25776232](https://our.internmc.facebook.com/intern/diff/D25776232/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25776232/)!

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/List.h:
```diff
@@ -53,6 +53,28 @@ bool operator==(const ListElementReference<T, Iterator>& lhs, const T& rhs);
+  using const_reference = const std::string&;
+};
+
+template<>
```
We could take this a bit further for all types with `std::is_reference_v<ListElementConstReferenceTraits<T>::const_reference>`, for example `optional<Tensor>` could benefit from this.

swolchok committed:
Update on "[PyTorch] List::operator[] can return const ref for Tensor & string"


This should supercede D21966183
(https://github.com/pytorch/pytorch/pull/39763) as the way to get fast
access to the contents of a `torch::List`.

Differential Revision: [D25776232](https://our.internmc.facebook.com/intern/diff/D25776232/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25776232/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] List::operator[] can return const ref for Tensor & string"


This should supercede D21966183
(https://github.com/pytorch/pytorch/pull/39763) as the way to get fast
access to the contents of a `torch::List`.

Differential Revision: [D25776232](https://our.internmc.facebook.com/intern/diff/D25776232/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25776232/)!

[ghstack-poisoned]
